Adds SHOULD_NOT_SLEEP to Destroy. Why didn't we do this before.

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Mostly sane INVOKE_ASYNC stuff, ~~outside of one UNLINT that I can't get around~~ mothblocks wuz here -Fuck you mothman
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

![image](https://user-images.githubusercontent.com/58055496/136157308-3cecec5d-11e5-4e5a-8aee-782bfdb4b548.png)

LemonInTheDark committed:
Adds SHOULD_NOT_SLEEP to Destroy. Why didn't we do this before.

Mostly sane INVOKE_ASYNC stuff. Nothing crazy

Mothblocks commented on code/modules/modular_computers/computers/item/computer.dm:
```diff
@@ -417,7 +417,7 @@
 	var/mob/user = usr
 	if(user && istype(user))
-		ui_interact(user) // Re-open the UI on this computer. It should show the main screen now.
+		UNLINT(INVOKE_ASYNC(src, /obj/item/modular_computer/ui_interact, user)) // Re-open the UI on this computer. It should show the main screen now.
```
What couldn't you figure out? Should at least be commented.

Mothblocks replied:
Like, I'm pretty sure you can just `/datum/proc/ui_interact` and you're fine.

LemonInTheDark replied:
byond doesn't like .proc/ui_interact, and it doesn't like /obj/item/modular_computer/proc/ui_interact, but smdmm doesn't like a proc path that goes nowhere

LemonInTheDark replied:
well uh

Ghilker commented on code/game/objects/structures/kitchen_spike.dm:
```diff
@@ -128,13 +128,13 @@
 	if(has_buckled_mobs())
-		for(var/mob/living/L in buckled_mobs)
-			release_mob(L)
+		for(var/mob/living/spiked_lad in buckled_mobs)
```
Look at you, cleaning up!

Cyberboss reviewed (commented):
Just checking, none of the `INVOKE_ASYNC`'d procs will fail due to the caller being deleted right? `malfvacate` in particular sounds sus to me.

LemonInTheDark commented:
> 
> 
> Just checking, none of the `INVOKE_ASYNC`'d procs will fail due to the caller being deleted right? `malfvacate` in particular sounds sus to me.

They would, but aren't invoke_async'd procs instantly called? they're async in the sense that they block sleeps from propagating up the chain, but I thought that was the limit of their effect

Cyberboss commented:
True, I meant after the sleeps

LemonInTheDark commented:
damn it I didn't consider that. Wow I'm stupid.

I'll do a pass then yeah. Guess the main issue is sleeping at all in destroy is asking to hang refs. I'll see about making sure they're all cleared away before it's done

LemonInTheDark committed:
Makes sure all but two destroys never sleep in such a way that would hang references to the object itself

LemonInTheDark committed:
as anything

LemonInTheDark committed:
whoops

LemonInTheDark committed:
explanations

LemonInTheDark reviewed (commented):
Some explanation on the two things I'm having issues with. One seems semi simple to solve, may bully kyler about the other, not sure yet

LemonInTheDark committed:
Fixes the lint errors on programs and computers.
In order to prevent sleeping in the destroying object, makes a subproc for use in computer/Destroy, and
invoke_async's the ui_act for the sake of programs

LemonInTheDark committed:
Whoops

LemonInTheDark committed:
Splits off just the "nuke old stuff, replace it with old map from load_program in such a way that we can avoid
sleeping in Destroy. Bit of a mess, paging kyler about it

LemonInTheDark committed:
begone

LemonInTheDark committed:
Moves clearing the past projection to its own proc

LemonInTheDark commented:
Kyler tells me it's alright
